### PR TITLE
feat(server): add ls archive support

### PR DESCRIPTION
**Pull request by stuckless**
Adds the ability such that when you double click on an archive file it sends back the archive listing instead of sending back the binary blob in the viewer.

image

There is a unit test for the archive.js, but I'm unaware of how the unit this in the server portion.

### CODE REVIEW

File: server/restafary.js [hunk 1]
```diff
@@ -190,11 +190,10 @@ async function sendArchiveListing(path, params) {
...
         let data = await archive.listArchive(path);
         let header = `Archive contents for ${path}`;
-        header += `\n${"=".repeat(header.length)}`;
+        header += `\n${'='.repeat(header.length)}`;
         data = `${header}\n${data}`;
         ponse.send(data, params);
-    } catch (ex) {
-        console.log(ex);
+    } catch {
         ponse.sendError('Unable to read archive contents for ' + path, params);
```
Comment (coderaiser) [inline server/restafary.js:197]: Let's move check into `listArchive`, and change it's signature into:

```js
const [error, list] = await listArchive(past);
```

Also let's create data in `archive.js` file, and here only send error or data.

File: server/archive.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+function isArchive(path) {
+    return lsArchive.isPathSupported(path);
+}
+
+function listArchive(path) {
+    return new Promise((resolve, error) => {
+        try {
+            lsArchive.list(path, (err, entries) => {
+                if (err) {
```
Comment (coderaiser) [inline server/archive.js:13]: I think there is no need in `try-catch` and `Promise` constructor. Let's use [util.promisify](https://nodejs.org/dist/latest-v15.x/docs/api/util.html#util_util_promisify_original) instead:

```js
const {promisify} = require('util');
const list = promisify(lsArhive.list);

module.exports.listArchive = (path) => {
    const [error, result] = await tryToCatch(list, path);
};
```

Let's use [try-to-catch](https://github.com/coderaiser/try-to-catch) here.

File: server/archive.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                    error(err);
+                    return;
+                }
+                const data = entries
+                    .map((z) => {
+                        return z.getPath();
+                    })
```
Comment (coderaiser) [inline server/archive.js:20]: Let's move out this function into `getPath` on top of the file (after require list):

```js
const getPath = (a) => a.getPath();
```

And use it with:

```js
const result = entries
    .map(getPath)
    .join('\n');

return result;
```

File: test/archive.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+const path = require('path');
+const test = require('supertape');
+const archive = require('../server/archive');
+
+test('archive: shows zip archive listing', async (t) => {
+    const file = path.resolve('./test/test.zip');
+    t.equal(archive.isArchive(file), true, 'Should be an archive file');
+    try {
+        const resp = await archive.listArchive(file);
+        t.equal(resp, `delete.js
```
Comment (coderaiser) [inline test/archive.js:12]: Let's use [montag](https://github.com/coderaiser/montag) here.

File: test/archive.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        t.fail(ex);
+    }
+    
+    t.end();
+});
+
+test('archive: returns false for non zip files', async (t) => {
+    const file = path.resolve('./test/not.a.zip.file');
+    t.equal(archive.isArchive(file), false, 'Should not be an archive file');
+    try {
```
Comment (coderaiser) [inline test/archive.js:32]: We can avoid using `try-catch` here if `listArchive` will contain one :).
Please use one assertion for one test.

File: server/restafary.js [hunk 6]
```diff
@@ -203,7 +216,13 @@ async function onGet(p, callback) {
             params.name = path;
         
         params.gzip = false;
-        ponse.sendFile(params);
+        
+        // Only process the archive listing if we are not given a query command
+        if (!p.request.originalUrl.includes('?') && archive.isArchive(path)) {
+            await sendArchiveListing(path, params);
```
Comment (coderaiser) [inline server/restafary.js:222]: I think you can simplify check to `if (p.query)`

**Review by coderaiser — CHANGES_REQUESTED** [review]

**Comment by coderaiser** [comment]
Nice, thank you :). Could you please tell me what is the problem with testing server?

Please fix lint errors with:
```
npm run fix:lint
```

**Comment by stuckless** [comment]
The lint errors are not mine, they are yours ;)  (ie, it's complaining that you are not using await on the `onGET` call, and since I didn't change that, i didn't fix it)

File: server/restafary.js (hunk 6 above)
Comment (stuckless) [reply server/restafary.js:222]: p.query always contains either the api path or the query command, so, I can't use a check on p.query.

**Comment by stuckless** [comment]
The problem with unit testing the server is that I don't know the tools that you are using to test the server, and they are not intuitive to me.

**Comment by coderaiser** [comment]
Please rebase master into your branch. And take a look at `test/get.js` for examples :).
There is two tools:
- [serve-once](https://github.com/coderaiser/node-serve-once)
- [mock-require](https://www.npmjs.com/search?q=mock-require)

**Comment by stuckless** [comment]
So, this is why I dislike contributing to other open source projects :(  There was nothing technically wrong with the code, but, I have to spend as much time changing it to fit into some programming style.  I had a couple of hours of spare time today to do this, and it's working on my server, which is all that I really wanted.  In the new few weeks, if I can find more time, I'll rework the code to try and make it more compliant with what you want, but I don't have that time in the next few weeks or so.  In all fairness, you could have written this code in less time than it has taken you to review it, and it would be perfectly in line with your style.

**Comment by coderaiser** [comment]
That's very sad, anyways thanks a lot for time you invested in developing this feature, and I hope you understand that I'm working on open source projects in my free time in the same way as you, and I don't really have it so much to implement all possible feature requests and fix all possible bugs, but I know that community of developers that can improve codebase this is what Cloud Commander really needs.

> There was nothing technically wrong with the code, but, I have to spend as much time changing it to fit into some programming style.

Look, this is technically not right to put `callback function` into try-catch block:

```js
try {
    lsArchive.list(path, (err, entries) => {
    });
} catch(ex) {
    error(ex); // how should reader understand that error is a function? why not reject?
}
```
It [want handle errors at all](https://www.joyent.com/node-js/production/design/errors).

Again, what purpose of this code:

```js
try {
    let data = await archive.listArchive(path);
    let header = `Archive contents for ${path}`;
    header += `\n${'='.repeat(header.length)}`;
    data = `${header}\n${data}`;
    ponse.send(data, params);
} catch {
    ponse.sendError('Unable to read archive contents for ' + path, params);
}
```

If there is already a library `try-to-catch`, which provide you much [more readable code](https://gist.github.com/coderaiser/a26e535bc43b5fe1ac4d72624bd6bed2) then usual try-catch block.
Problem with this block is lacking of [separation concern](https://www.castsoftware.com/blog/how-to-implement-design-pattern-separation-of-concerns#:~:text=Separation%20of%20concerns%20is%20a,of%20concerns%20is%20about%20order.).

What I want mention is, what I'm talking about it's not code style, this is a basis of JavaScript Development. And everybody will win if you became a little bit more qualified developer after this Pull Request. Also you need to understand that I should maintain this code after you leave: fix bugs, security vulnerabilities, add new features. 

I understand that can be sad to read points on code review, but this is much better, then PR that receives no answers at all, which I have a lot :), so I'm usually very grateful for any comment people give me.

**Comment by stuckless** [comment]
@coderaiser I respect your position as the maintainer.  I maintain many projects myself, and something that I've learned in my long career that newbies eventually have to figure out is that there is a difference between code that is wrong and code that is different than I would write.  I do about 10 code reviews a day and I see lots of code that is very different than I would write, but I still allow it, provided code is not flawed.  Granted when I first started out, I did a lot of "this is not how I'd do it, so it's wrong" and so I rejected it.

> If there is already a library try-to-catch, which provide you much more readable code then usual try-catch block.

Again, more readable to who?  try/catch is an industry documented pattern and you are suggesting to not use that but instead use some homegrown function because it's more readable to you (maybe).  But I'm guessing that even you understand the try/catch block, no?  So this boils down to personal preference, not a real problem.  You can't possibly believe that the world is going to know and understand your homegrown try/catch obfuscation more than the industry documented try/catch pattern, do you?

Are you asking what is the purpose of the try/catch around an async/await block.  Again, this async/await has a documented pattern of returning a response OR throwing the error. So, I'm not sure how to answer your question on that, since, given how you've gone out of your way to obfuscate promises in your code, I'm guessing you fully understand how they work?

> error(ex); // how should reader understand that error is a function? why not reject?

Not sure what this question is?  Again, it's part of the documented Promise contract.  So, yeah, I expect that the reader will either know how promises work, or they will go to the public resources on Promises and understand that the second part of a promise is the function to call when there is an error and you want to reject and notify about the error?  Is that what you are asking?  Here's a link to the Promise documentation if that helps you understand them.  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

> What I want mention is, what I'm talking about it's not code style, this is a basis of JavaScript Development. And everybody will win if you became a little bit more qualified developer after this Pull Request. Also you need to understand that I should maintain this code after you leave: fix bugs, security vulnerabilities, add new features.

And that I fully get.  You need to maintain this code, so you are asking that the style "look" like you wrote the code. Which was my original comment.  The challenge with trying to help newer maintainers with their projects is a frustrating one, because, much of the coding is not about providing a feature, but rather, ensuring that you are can write code that makes it look like the maintainer wrote the code because every maintainer has a set of personal preferences like "try-with-catch is more readable to me" or "please create obfuscated functions that return arrays because that's more readable to me".

You have a good project here, and I'm glad you built it.  My goal was to help you enhance it, and make it better for everyone else (the open-source dream that is rarely realized).  This was a fun diversion for a couple of hours away from my real job.  This has to be fun. If it feels like more work than real work, then I'm not interested.

I'll close the PR maybe someday, someone will contribute the archive viewer in your code style.  Until then, I have my own fork, and this is running on my server and is tremendously helpful (I used it about 10 times already yesterday).  Good luck, I really do wish you well, because this is a great tool.

**Pull request closed without merging**